### PR TITLE
[IMP] sale: auto-confirm & auto-invoice payment flows

### PR DESCRIPTION
**Pull request by chevalierv**
Rework and code cleanup of automatic invoice and auto-confirm order flows.

task-2541188

### CODE REVIEW

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -849,31 +849,40 @@ def _show_cancel_wizard(self):
...
+        if force_confirmation_template or (
+            all(order.state == 'sale' for order in self) and not self.env.context.get('proforma')
+        ):
+            default_template_id = int(
+                self.env['ir.config_parameter'].sudo()
+                                               .get_param('sale.default_confirmation_template')
+            )
+            default_template = self.env['mail.template'].browse(default_template_id)
+            return default_template or self.env.ref('sale.mail_template_sale_confirmation')
+        return self.env.ref('sale.email_template_edi_sale')
```
Comment (chevalierv) [inline addons/sale/models/sale_order.py:872]: @Demesmaeker Can you review this change? I added back an improvement that we've made in our first task #68965.

Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:872]: Ok for me, maybe add _sudo in the variable name, as the fact that it's sudo will stay (it's a tip Antoine gave me some days ago). It's better to put it in the name, so we don't forget about it.

File: addons/sale/models/sale_order.py [hunk 2]
```diff
@@ -849,31 +849,40 @@ def _show_cancel_wizard(self):
...
+        :rtype: record of `mail.template`
+        """
+        if force_confirmation_template or (
+            all(order.state == 'sale' for order in self) and not self.env.context.get('proforma')
+        ):
+            _sudo_default_template_id = int(
+                self.env['ir.config_parameter'].sudo()
+                                               .get_param('sale.default_confirmation_template')
+            )
+            _sudo_default_template = self.env['mail.template'].browse(_sudo_default_template_id)
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:870]: Since it is a template id, it's an integer, and therefore doesn't have any notion of sudo or not ;)
```suggestion
            default_template_id = int(
                self.env['ir.config_parameter'].sudo()
                                               .get_param('sale.default_confirmation_template')
            )
            default_template = self.env['mail.template'].browse(default_template_id)
```
Furthermore, your _sudo_default_template is not sudoed, since you're only browsing an id, it's confusing to include a sudo in its name.

Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:870]: My bad, I saw "sudo" and didn't put more thought in it. I'll keep it in mind for next time.

File: addons/sale/models/payment_transaction.py [hunk 3]
```diff
@@ -46,34 +53,37 @@ def _set_pending(self, state_message=None):
...
-                _logger.warning(
-                    '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %r, got %r',
-                    self.acquirer_id.provider,order.name, order.id,
-                    order.amount_total, self.amount,
+        """ Auto-confirm sales orders based on the amount of all the transactions linked. """
+        for tx in self:
+            for order in tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent')):
+                tx_total_amount = math.fsum(  # Sum the amount of each transaction in the SO
+                    order.transaction_ids.mapped(lambda tx: float(tx.amount))
                 )
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:61]: Missing check on currencies, what if the tx currency is different than the order currency ? (Not your fault, it wasn't verified in old code).

File: addons/sale/models/sale_order.py [hunk 4]
```diff
@@ -849,31 +847,44 @@ def _show_cancel_wizard(self):
...
         ''' Opens a wizard to compose an email, with relevant mail template loaded by default '''
         self.ensure_one()
-        template_id = self._find_mail_template()
+        mail_template = self._find_mail_template()
         lang = self.env.context.get('lang')
-        template = self.env['mail.template'].browse(template_id)
-        if template.lang:
-            lang = template._render_lang(self.ids)[self.id]
+        if mail_template.lang:
+            lang = mail_template._render_lang(self.ids)[self.id]
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:891]: Modified behavior:
Previously, `_find_mail_template` returned an id (or False if not found), then browsing it gave an empty recordset (which doesn't raise if you access its fields)

Now, it returns a recordset or None if not found, which will raise if the records are not found.

Or you should still be able to open the "send mail" wizard even if all the templates have been deleted.

File: addons/sale/tests/test_payment_flow.py [hunk 5]
```diff
@@ -66,6 +66,7 @@ def test_11_so_payment_link(self):
         with mute_logger('odoo.addons.payment.models.payment_transaction'):
             processing_values = self.get_processing_values(**route_values)
         tx_sudo = self._get_tx(processing_values['reference'])
+        self.order.transaction_ids.write({Command.link(tx_sudo.id)})
```
Comment (Feyensv) [inline addons/sale/tests/test_payment_flow.py]: It was a cache problem, this line can be removed.

**Review by Feyensv — CHANGES_REQUESTED** [review]
Quick comments noticed while fixing the test.

File: addons/sale/models/sale_order.py [hunk 6]
```diff
@@ -879,21 +879,21 @@ def _find_mail_template(self, force_confirmation_template=False):
             if default_confirmation_template:
                 return default_confirmation_template
             else:
-                return self.env.ref('sale.mail_template_sale_confirmation')
-        return self.env.ref('sale.email_template_edi_sale')
+                return self.env.ref('sale.mail_template_sale_confirmation', None)
+        return self.env.ref('sale.email_template_edi_sale', None)
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:883]: `raise_if_not_found=None` ?

Comment (chevalierv) [reply addons/sale/models/sale_order.py:883]: I have obviously mistaken this parameter. I thought that it returned the mail.template or None. I'll change that right away.

File: addons/sale/models/payment_transaction.py (hunk 3 above)
Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:51]: I didn't know that was possible. In that case, it means that I need to convert on an arbitrary currency using the current rate? But what if the current rate change between the moment when the transaction is made and the moment that the post-process is made? Do we have a history of rates so I can fetch the current rate at the time when the transaction was paid?

Comment (Feyensv) [reply addons/sale/models/payment_transaction.py:51]: It's a tricky case that may happen (but shouldn't indeed :D).
About currency rates, there exists tools to compute rates at a given date (and records to store such info), but I don't think you should go that far.

If the transaction currencies are different, you should skip SO validation IMHO, we don't want to support that case (I'm not even sure we want to support the confirmation of the SO after multiple partial payments ^^).

File: addons/sale/models/payment_transaction.py [hunk 7]
```diff
@@ -89,36 +84,39 @@ def _log_message_on_linked_documents(self, message):
...
-            if default_template:
-                for trans in self.filtered(lambda t: t.sale_order_ids.filtered(lambda so: so.state in ('sale', 'done'))):
-                    trans = trans.with_company(trans.acquirer_id.company_id).with_context(
-                        mark_invoice_as_sent=True,
-                        company_id=trans.acquirer_id.company_id.id,
-                    )
-                    for invoice in trans.invoice_ids.with_user(SUPERUSER_ID):
-                        invoice.message_post_with_template(int(default_template), email_layout_xmlid="mail.mail_notification_paynow")
-        return res
+        self.mapped('sale_order_ids').filtered(
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:89]:
```suggestion
        self.sale_order_ids.filtered(
```
mapped is not necessary for relational fields

File: addons/sale/models/sale_order.py [hunk 8]
```diff
@@ -952,10 +966,10 @@ def _send_order_confirmation_mail(self):
         if self.env.su:
             # sending mail in sudo was meant for it being sent from superuser
             self = self.with_user(SUPERUSER_ID)
-        template_id = self._find_mail_template(force_confirmation_template=True)
-        if template_id:
+        mail_template = self._find_mail_template(force_confirmation_template=True)
+        if mail_template:
             for order in self:
-                order.with_context(force_send=True).message_post_with_template(template_id, composition_mode='comment', email_layout_xmlid="mail.mail_notification_paynow")
+                order.with_context(force_send=True).message_post_with_template(mail_template.id, composition_mode='comment', email_layout_xmlid="mail.mail_notification_paynow")
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:970]: You can batch the `message_post_with_template` call as well (for performance purposes)

File: addons/sale/models/sale_order.py [hunk 9]
```diff
@@ -892,31 +892,45 @@ def _show_cancel_wizard(self):
...
+                IrConfigParam.get_param('sale.default_confirmation_template')
+            )
+            default_confirmation_template = self.env['mail.template'].browse(
+                default_confirmation_template_id
+            ).exists()
+            if default_confirmation_template:
+                return default_confirmation_template
+            else:
+                return self.env.ref('sale.mail_template_sale_confirmation', False)
+        return self.env.ref('sale.email_template_edi_sale', False)
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:918]: I'd prefer one return and a variable set in if..else branches :)

**Review by Feyensv — COMMENTED** [review]
In `_set_authorized` and `_reconcile_after_done`, If you pay for a confirmed SO, I believe an additional confirmation mail will still be sent to the customer.

Two solutions IMHO:
* `_check_amount_and_confirm_order` could return the confirmed SO's for which a confirmation email must be sent
* `_send_order_confirmation_mail` could be moved after `action_confirm` in `_check_amount_and_confirm_order`

File: addons/sale/models/sale_order.py (hunk 8 above)
Comment (chevalierv) [reply addons/sale/models/sale_order.py:970]: I don't understand what I should do here. Can you give me an example?

Comment (Feyensv) [reply addons/sale/models/sale_order.py:970]: `self.with_context(force_send=True).message_post_with_template(...)`, the for loop (and multi recordset consideration) can be delegated to `message_post_with_template` bc it supports calls with multiple records in self.

**Comment by chevalierv** [comment]
> In `_set_authorized` and `_reconcile_after_done`, If you pay for a confirmed SO, I believe an additional confirmation mail will still be sent to the customer.
> 
> Two solutions IMHO:
> 
> `_check_amount_and_confirm_order` could return the confirmed SO's for which a confirmation email must be sent
> `_send_order_confirmation_mail` could be moved after `action_confirm` in `_check_amount_and_confirm_order`
> 

I see your point. It is probably a mistake from the many times that I've changed the way we send mail in this task. IMO, `_set_authorized` is already okay because there is a filter to ensure to consider only SO in draft or sent state. I've adapted `_reconcile_after_done` following your first advice.

File: addons/sale/models/sale_order.py (hunk 8 above)
Comment (chevalierv) [reply addons/sale/models/sale_order.py]: I tried your solution but I've got red run bots. I rollbacked to the unbatch solution for now.

File: addons/sale/models/payment_transaction.py [hunk 10]
```diff
@@ -49,32 +46,37 @@ def _set_pending(self, state_message=None):
...
     def _set_authorized(self, state_message=None):
         """ Override of payment to confirm the quotations automatically. """
         super()._set_authorized(state_message=state_message)
         sales_orders = self.mapped('sale_order_ids').filtered(lambda so: so.state in ('draft', 'sent'))
-        for tx in self:
-            tx._check_amount_and_confirm_order()
-
+        self._check_amount_and_confirm_order()
         # send order confirmation mail
         sales_orders._send_order_confirmation_mail()
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:81]:
```suggestion
        confirmed_orders = self._check_amount_and_confirm_order()
        # send order confirmation mail
        confirmed_orders._send_order_confirmation_mail()
```
(and remove `sales_orders` variable)

Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:81]: @AntoineVDV was clear that we should not change the way sending mail works in this task. If the mail is not sent twice, it's not a bug. Although, I agree that the mailing system is wobbly and should be improved.

Comment (Feyensv) [reply addons/sale/models/payment_transaction.py:81]: image
(on your branch, with a little tweak to support authorization with "Test")

Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:81]: Shit. I though I was getting away with this.I'll look into it tomorrow :p

File: addons/sale/models/payment_transaction.py [hunk 11]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
-                order.message_post(
-                    subject=_("Amount Mismatch (%s)", self.acquirer_id.provider),
-                    body=_("The order was not confirmed despite response from the acquirer (%s): order total is %r but acquirer replied with %r.") % (
-                        self.acquirer_id.provider,
-                        order.amount_total,
-                        self.amount,
+        """ Confirm the sale order based on the amount of a transaction.
+
+        :return: confirmed sale orders
+        :rtype: a set of sale orders
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:52]:
```suggestion
        :rtype: a `sale.order` recordset
```

File: addons/sale/models/payment_transaction.py [hunk 12]
```diff
@@ -92,33 +93,36 @@ def _log_message_on_linked_documents(self, message):
...
-        self._invoice_sale_orders()
-        res = super()._reconcile_after_done()
-        if self.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice') and any(so.state in ('sale', 'done') for so in self.sale_order_ids):
-            self.filtered(lambda t: t.sale_order_ids.filtered(lambda so: so.state in ('sale', 'done')))._send_invoice()
-        return res
+        confirmed_orders._send_order_confirmation_mail()
+
+        # invoice the sale orders if needed and send it
+        if self.env['ir.config_parameter'].sudo().get_param('sale.automatic_invoice'):
+            self._invoice_sale_orders()
```
Comment (Feyensv) [inline addons/sale/models/payment_transaction.py:102]: Little comment to explain why we do not reuse `confirmed_orders` here (bc of partial payments & co).

File: addons/sale/models/sale_order.py [hunk 13]
```diff
@@ -699,17 +698,33 @@ def action_quotation_send(self):
...
+        If all SOs are 'sale' or ``force_confirmation_template`` is `True`, the mail template is
+        that of the sale confirmation. Else, it returns the quotation's email template.
 
-        return template_id
+        :param bool force_confirmation_template: Allows to force the confirmation template to be
+        use in any case
+        :return: The correct mail template based on the current status
+        :rtype: record of `mail.template` or `None` if not found
+        """
+        return_template = self.env.ref('sale.email_template_edi_sale', False)
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:711]: could be moved in an else clause, not used if we enter the if.
(or another if to fallback when `mail_template_sale_confirmation` is not found, to keep old behavior)

File: addons/sale/models/sale_order.py [hunk 14]
```diff
@@ -764,10 +779,13 @@ def _send_order_confirmation_mail(self):
         if self.env.su:
             # sending mail in sudo was meant for it being sent from superuser
             self = self.with_user(SUPERUSER_ID)
-        template_id = self._find_mail_template(force_confirmation_template=True)
-        if template_id:
-            for order in self:
-                order.with_context(force_send=True).message_post_with_template(template_id, composition_mode='comment', email_layout_xmlid="mail.mail_notification_paynow")
+        mail_template = self._find_mail_template(force_confirmation_template=True)
+        if mail_template and len(self) > 0:
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:783]:
```suggestion
        if mail_template:
```

the self check should be moved in the beginning of the method:
```python
if not self:
    return
```

(and btw `len(self) > 0 == bool(self)`)

**Review by Feyensv — COMMENTED** [review]
Rebase needed :grinning:

**Review by Feyensv — APPROVED** [review]

File: addons/sale/models/sale_order.py [hunk 15]
```diff
@@ -726,17 +725,34 @@ def action_quotation_send(self):
...
+        ):
+            IrConfigParam = self.env['ir.config_parameter'].sudo()
+            default_confirmation_template_id = int(
+                IrConfigParam.get_param('sale.default_confirmation_template')
+            )
+            default_confirmation_template = self.env['mail.template'].browse(
+                default_confirmation_template_id
+            ).exists()
+            if default_confirmation_template:
+                return_template = default_confirmation_template
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:750]: Weird name for a variable. Since you're at the very end of the method, you could alternatively return directly.

File: addons/sale/models/sale_order.py [hunk 16]
```diff
@@ -726,17 +725,34 @@ def action_quotation_send(self):
...
+            else:
+                return_template = self.env.ref('sale.mail_template_sale_confirmation', False)
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:752]: explicit > implicit

Especially here where one could assume `False` is the fallback value of `ref` (like in `.get()`). Also, it prevents bad jokes when the method is refactored.
```suggestion
                return_template = self.env.ref('sale.mail_template_sale_confirmation', raise_if_not_found=False)
```

File: addons/sale/models/sale_order.py [hunk 17]
```diff
@@ -790,13 +806,20 @@ def _get_forbidden_state_confirm(self):
...
-        if template_id:
-            for order in self:
-                order.with_context(force_send=True).message_post_with_template(template_id, composition_mode='comment', email_layout_xmlid="mail.mail_notification_paynow")
+
+        mail_template = self._find_mail_template(force_confirmation_template=True)
+        if mail_template:
+            self.with_context(force_send=True).message_post_with_template(
+                mail_template.id,
+                composition_mode='comment',
+                email_layout_xmlid="mail.mail_notification_paynow"
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:821]: Consistency
```suggestion
                composition_mode='comment',
                email_layout_xmlid='mail.mail_notification_paynow'
```

Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:821]: Nitpicking.
```suggestion
                composition_mode='comment',
                email_layout_xmlid="mail.mail_notification_paynow",
```

File: addons/sale/models/payment_transaction.py [hunk 18]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
-                    self.acquirer_id.provider,order.name, order.id,
-                    order.amount_total, self.amount,
-                )
-                order.message_post(
-                    subject=_("Amount Mismatch (%s)", self.acquirer_id.provider),
-                    body=_("The order was not confirmed despite response from the acquirer (%s): order total is %r but acquirer replied with %r.") % (
-                        self.acquirer_id.provider,
-                        order.amount_total,
-                        self.amount,
+        """ Confirm the sale order based on the amount of a transaction.
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:49]:
```suggestion
        """ Confirm the sales order based on the amount of a transaction.
```
This is enough as the title of the docstring but there should at least be some information on the heuristic used to confirm the quotation in the body of the docstring.

File: addons/sale/models/payment_transaction.py [hunk 19]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
+
+        :return: confirmed sale orders
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:51]: sale**s** order
```suggestion
        :return: The confirmed sales orders.
```

File: addons/sale/models/payment_transaction.py [hunk 20]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
 
     def _set_authorized(self, state_message=None):
         """ Override of payment to confirm the quotations automatically. """
         super()._set_authorized(state_message=state_message)
-        sales_orders = self.mapped('sale_order_ids').filtered(lambda so: so.state in ('draft', 'sent'))
-        for tx in self:
-            tx._check_amount_and_confirm_order()
-
+        confirmed_orders = self._check_amount_and_confirm_order()
         # send order confirmation mail
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:79]: Unnecessary comment.

File: addons/sale/models/payment_transaction.py [hunk 21]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
-                        self.amount,
+        """ Confirm the sale order based on the amount of a transaction.
+
+        :return: confirmed sale orders
+        :rtype: a `sale.order` recordset
+        """
+        confirmed_orders = self.env['sale.order']
+        for tx in self:
+            # We only support the flow where exactly one quotation is linked to a transaction and
+            # vice versa.
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:57]: This could be for another task but we should at least support the case where multiple transactions are linked to a quotation (partial payment). cc @Feyensv @dalaOdoo

File: addons/sale/models/payment_transaction.py [hunk 22]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
+            order = tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent'))
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:58]: Plural, if we don't support handle mutliple quotations (orders) at once.
```suggestion
            quotations = tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent'))
```

File: addons/sale/models/payment_transaction.py [hunk 23]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
+            if len(tx.sale_order_ids) == 1 and order and len(order.transaction_ids) == 1:
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:59]: Couldn't we just check `len(quotations) == 1`?

File: addons/sale/models/payment_transaction.py [hunk 24]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
+                    confirmed_orders |= order
+                else:
+                    _logger.warning(
+                        '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %s, got %s',
+                        tx.provider,
+                        order.name,
+                        order.id,
+                        format_amount(order.env, order.amount_total, order.currency_id),
+                        format_amount(tx.env, tx.amount, tx.currency_id),
                     )
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:72]: We lost the `message_post` along the way. Is that intended? I don't see it mentioned in the spec. cc @dalaodoo

File: addons/sale/models/payment_transaction.py [hunk 25]
```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
...
+                    order.with_context(send_email=True).action_confirm()
+                    confirmed_orders |= order
+                else:
+                    _logger.warning(
+                        '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %s, got %s',
+                        tx.provider,
+                        order.name,
+                        order.id,
+                        format_amount(order.env, order.amount_total, order.currency_id),
+                        format_amount(tx.env, tx.amount, tx.currency_id),
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:71]: That's %s a %s lot %s of %s parameters. It'd be clearer if you use a `dict` for the interpolation.

File: addons/sale/models/payment_transaction.py [hunk 26]
```diff
@@ -92,33 +93,38 @@ def _log_message_on_linked_documents(self, message):
 
     def _reconcile_after_done(self):
         """ Override of payment to automatically confirm quotations and generate invoices. """
-        sales_orders = self.mapped('sale_order_ids').filtered(lambda so: so.state in ('draft', 'sent'))
-        for tx in self:
-            tx._check_amount_and_confirm_order()
+        confirmed_orders = self._check_amount_and_confirm_order()
         # send order confirmation mail
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:97]: Unnecessary comment.

File: addons/sale/models/payment_transaction.py [hunk 27]
```diff
@@ -92,33 +93,38 @@ def _log_message_on_linked_documents(self, message):
...
+            'sale.default_invoice_email_template'
+        )
+        if not template_id:
             return
 
-        for trans in self:
-            trans = trans.with_company(trans.acquirer_id.company_id).with_context(
-                company_id=trans.acquirer_id.company_id.id,
+        for tx in self:
+            tx = tx.with_company(tx.acquirer_id.company_id).with_context(
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:116]:
```suggestion
            tx = tx.with_company(tx.company_id).with_context(
```

File: addons/sale/models/payment_transaction.py [hunk 28]
```diff
@@ -92,33 +93,38 @@ def _log_message_on_linked_documents(self, message):
...
-            invoice_to_send = trans.invoice_ids.filtered(
+            invoice_to_send = tx.invoice_ids.filtered(
                 lambda i: not i.is_move_sent and i.state == 'posted' and i._is_ready_to_be_sent()
             )
             invoice_to_send.is_move_sent = True # Mark invoice as sent
             for invoice in invoice_to_send.with_user(SUPERUSER_ID):
-                invoice.message_post_with_template(int(default_template), email_layout_xmlid="mail.mail_notification_paynow")
+                invoice.message_post_with_template(
+                    int(template_id),
+                    email_layout_xmlid="mail.mail_notification_paynow"
```
Comment (AntoineVDV) [inline addons/sale/models/payment_transaction.py:126]: Nitpicking.
```suggestion
                    email_layout_xmlid='mail.mail_notification_paynow',
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
I posted a lot of comments (mostly typos and nitpicking) but the overall structure of the flow is already a lot better than it was :+1:

File: addons/sale/models/sale_order.py [hunk 29]
```diff
@@ -1052,7 +1056,7 @@ def _create_invoices(self, grouped=False, final=False, date=None):
             invoice_vals['invoice_line_ids'] += invoice_line_vals
             invoice_vals_list.append(invoice_vals)
 
-        if not invoice_vals_list:
+        if not (invoice_vals_list or self._context.get('do_not_raise')):
```
[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/sale/models/sale_order.py]: I agree with you but I'll need some help on the case where we want to raise an UserError. I think that it is only when we create an invoice from the wizard, so I changed it. Do you think of any other cases? cc @Feyensv

[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/models/sale_order.py]: We do not want to change this behavior here (and that's why the runbot is red).
Keep the UserError by default.

File: addons/sale/models/payment_transaction.py (hunk 22 above)
Comment (AntoineVDV) [reply addons/sale/models/payment_transaction.py]: Then it should probably be renamed between here and L70-79 because it might still be the case that `quotation` holds multiple records, no? Even if we don't support a given flow, we should name our variables according to what they (might be) are.

File: addons/website_sale/models/website.py [hunk 30]
```diff
@@ -418,7 +418,7 @@ def _get_current_pricelist_id(self, partner_sudo):
 
     @api.model
     def _get_current_fiscal_position_id(self, partner_sudo):
-        AccountFiscalPosition = self.env['account.fiscal.position']
+        AccountFiscalPosition = self.env['account.fiscal.position'].sudo()
```
Comment (Feyensv) [inline addons/website_sale/models/website.py:421]: This file diff is unrelated (and already fixed in master anyway) :eyes:

Comment (chevalierv) [reply addons/website_sale/models/website.py]: I told you that I changed the files :p Now I see why I didn't find it after...

File: addons/sale/models/sale_order.py [hunk 31]
```diff
@@ -737,17 +736,31 @@ def action_quotation_send(self):
...
+        """
+        if force_confirmation_template or (
+            all(order.state == 'sale' for order in self) and not self.env.context.get('proforma')
+        ):
+            default_confirmation_template_id = self.env['ir.config_parameter'].sudo().get_param(
+                'sale.default_confirmation_template', 0
+            )
+            default_confirmation_template = self.env['mail.template'].browse(
+                default_confirmation_template_id
+            ).exists()
```
Comment (AntoineVDV) [inline addons/sale/models/sale_order.py:757]: The potentially unnecessary browse reported in https://github.com/odoo/odoo/pull/72368#discussion_r895490824 is still here.
```suggestion
            default_confirmation_template_id = self.env['ir.config_parameter'].sudo().get_param(
                'sale.default_confirmation_template'
            )
            default_confirmation_template = default_confirmation_template_id and self.env['mail.template'].browse(
                default_confirmation_template_id
            ).exists()
```

File: addons/sale/models/payment_transaction.py (hunk 21 above)
Comment (AntoineVDV) [reply addons/sale/models/payment_transaction.py:61]: @chevalierv Can you please create the feedback and close this comment?

Comment (chevalierv) [reply addons/sale/models/payment_transaction.py:61]: Done.

Feedback: task-2892344

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Comment by AntoineVDV** [comment]
@robodoo rebase-ff
